compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/replacements/CounterModeSubstitutions.java

```diff
@@ -0,0 +1,84 @@
+import static org.graalvm.compiler.replacements.ReplacementsUtil.getArrayBaseOffset;
+
+import org.graalvm.compiler.api.replacements.ClassSubstitution;
+import org.graalvm.compiler.api.replacements.MethodSubstitution;
+import org.graalvm.compiler.debug.DebugContext;
+import org.graalvm.compiler.debug.GraalError;
+import org.graalvm.compiler.hotspot.HotSpotBackend;
+import org.graalvm.compiler.nodes.ComputeObjectAddressNode;
+import org.graalvm.compiler.nodes.PiNode;
+import org.graalvm.compiler.nodes.extended.RawLoadNode;
+import org.graalvm.compiler.replacements.ReplacementsUtil;
+import org.graalvm.compiler.word.Word;
+import org.graalvm.word.LocationIdentity;
+import org.graalvm.word.Pointer;
+import org.graalvm.word.WordFactory;
+
+import jdk.vm.ci.meta.JavaKind;
+import sun.misc.Unsafe;
+
+@ClassSubstitution(className = "com.sun.crypto.provider.CounterMode", optional = true)
+public class CounterModeSubstitutions {
+
+    @MethodSubstitution(isStatic = false)
+    static int implCrypt(Object receiver, byte[] in, int inOff, int len, byte[] out, int outOff) {
+        Object realReceiver = piCastNonNull(receiver, HotSpotReplacementsUtil.methodHolderClass(INJECTED_INTRINSIC_CONTEXT));
+        Object embeddedCipher = RawLoadNode.load(realReceiver, embeddedCipherOffset(INJECTED_INTRINSIC_CONTEXT), JavaKind.Object, LocationIdentity.any());
+        Object aesCipher = piCastNonNull(embeddedCipher, aesCryptType(INJECTED_INTRINSIC_CONTEXT));
+
+        Word srcAddr = WordFactory.unsigned(ComputeObjectAddressNode.get(in, ReplacementsUtil.getArrayBaseOffset(INJECTED_METAACCESS, JavaKind.Byte) + inOff));
+        Word dstAddr = WordFactory.unsigned(ComputeObjectAddressNode.get(out, ReplacementsUtil.getArrayBaseOffset(INJECTED_METAACCESS, JavaKind.Byte) + outOff));
+
+        int cntOffset = HotSpotReplacementsUtil.getFieldOffset(
+                        HotSpotReplacementsUtil.getType(INJECTED_INTRINSIC_CONTEXT, "Lcom/sun/crypto/provider/CounterMode;"),
+                        "counter");
+        int encCntOffset = HotSpotReplacementsUtil.getFieldOffset(
+                        HotSpotReplacementsUtil.getType(INJECTED_INTRINSIC_CONTEXT, "Lcom/sun/crypto/provider/CounterMode;"),
+                        "encryptedCounter");
+        int usedOffset = HotSpotReplacementsUtil.getFieldOffset(
+                        HotSpotReplacementsUtil.getType(INJECTED_INTRINSIC_CONTEXT, "Lcom/sun/crypto/provider/CounterMode;"),
+                        "used");
+        Object kObject = RawLoadNode.load(aesCipher, AESCryptSubstitutions.kOffset(INJECTED_INTRINSIC_CONTEXT), JavaKind.Object, LocationIdentity.any());
+        Object cntObj = RawLoadNode.load(realReceiver, cntOffset, JavaKind.Object, LocationIdentity.any());
+        Object encCntObj = RawLoadNode.load(realReceiver, encCntOffset, JavaKind.Object, LocationIdentity.any());
+        Object used = RawLoadNode.load(realReceiver, usedOffset, JavaKind.Int, LocationIdentity.any());
```

Comment:
This is not an object.

Reply:
Thanks for the review Yudi. Let me fix these things. The above change could be used throughout the rest of the method substitutions as well I think.

Reply:
@mur47x111 Just looked the type for used. I could not find another load type of Integer. What should the type of used be in this case if not object?

Reply:
You could create one:
```
    @NodeIntrinsic(value = RawLoadNode.class)
    private static native int loadIntFromObjectIntrinsic(Object object, long offset, @ConstantNodeParameter JavaKind wordKind, @ConstantNodeParameter LocationIdentity locationIdentity);
```

Reply:
Sorry Yudi, but which native call does this correspond to? I cannot find the source for the native calls such as loadWordFromObjectIntrinsic.

Reply:
There are no implementations to `@NodeIntrinsic`-annotated methods. Graal will create IR nodes (of the type specified by `value=`) when encountering invocations to these methods.

Reply:
But how to RawLoadNode can figure out how to return an int?

Reply:
For example, it depends on the `JavaKind` parameter that we feed to `loadWordFromObjectIntrinsic`. You could also take a look at the generated source code in Plugin_HotSpotReplacementsUtil.java. You might look for the class `Plugin_HotSpotReplacementsUtil_loadWordFromObjectIntrinsic`.

Anyway, I think we don't need to load the value here. Just calculate the address of `used` might be sufficient. I.e.,
``
        Word usedPtr = WordFactory.unsigned(ComputeObjectAddressNode.get(realReceiver, usedOffset(INJECTED_INTRINSIC_CONTEXT)));
```